### ISSUE TITLE
eachTuple performance / Cache tuples

### ISSUE THREAD

**Issue by rpbouman** [rpbouman/xmla4js#19]
eachTuple is a convenient way to go through the tuples of a dataset axis. In a typical example, one will process the tuples from "high" axes toword "low" axes. 

For example: in a 2 dimensional result, a typical way to represent it would be a table. We would first take the row axis, and for each tuple, iterate through the tuples of the column axis, and then render the cell that corresponds to the intersection of row and column tuples.

If this iteration is done with eachTuple, a tuple will be read from the reponseXml. However, in the majority of cases, the tuple itself will be required only once to render a header. The repeated iteration over the lower axes (i.e., for each row, go through all the columns) is required only to access and render a cell. So in the majority of cases, we're doing spurious work. This has a performance overhead which can rapidly add up as we have more axes and more tuples.

The current work around I use is to only use eachTuple when I actually need the tuples, maybe cache the tuples in an array if I need them a second time around, and do all remaining iteration useing a straight loop from 0 through axis.tupleCount(). This works and it saves time but it has the drawback that our code uses 2 different types of axis processing. We're doing it for good reasons and also for functionally different cases but it feels as if its unnecessarily complex.

If eachTuple would itself create an array to cache the tuples it would then be able to extract tuples from that cache in subsequent iterations of that axis. This would at least save the time extracting tuple data from the response xml and have little impact on the first iteration (it would cost a little bit of memory)

**Comment by babnik63** [comment]
Nice;)

**Comment by rpbouman** [comment]
Hi @babnik63, is this currently a problem you run into? It isn't currently affecting me since I already wrote everything using my workaround. If you are experiencing a problem, then I would really appreciate feedback on how you'd like this improvement to work.

The only thing I do not want to do is break backward compatibility, but otherwise I'm open-minded.

My initial thought is to have a flag in the axis configuration, "cacheTuples", which would default to true, or maybe default to true for all axes except for the Slicer and the "highest" axis. And of course there would be a getter/setter to control this option.

**Comment by rpbouman** [comment]
Closing as won't fix. Applications can always fetch the axis as object, and use standard javascript array methods to iterate through the axis' tuples.
